### PR TITLE
Add support for WebAssembly Micro Runtime (WAMR).

### PR DESCRIPTION
**Pull request by leyao-daily**
Add WAMR support

### CODE REVIEW

**Comment by leyao-daily** [comment]
@googlebot I signed it!

**Comment by leyao-daily** [comment]
@googlebot I signed it!

**Comment by leyao-daily** [comment]
@googlebot I fixed it.

**Comment by PiotrSikora** [comment]
> @googlebot I fixed it.

One of your commits has a line saying `Co-authored-by: Liang He <liang.he@intel.com>`, and we don't have signed CLA for that person.

**Comment by leyao-daily** [comment]
> > @googlebot I fixed it.
> 
> @leyao-daily one of your commits has a line saying `Co-authored-by: Liang He <liang.he@intel.com>`, and we don't have signed CLA for that person.

Okay, I will remind him.

**Comment by mathetake** [comment]
I believe you need to put
```cpp
 #if defined(WASM_WAMR) 
     } else if (runtime_ == "wamr") { 
       vm_ = proxy_wasm::createWAMRVm(); 
 #endif 
```
[here](https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/d06e7f9a65a5337995cdcd4765814749fd3f0205/test/utility.h#L67-L85). 

and
```cpp
#if defined(WASM_WAMR)
    "wasmtime",
#endif
```
[here](https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/d06e7f9a65a5337995cdcd4765814749fd3f0205/test/utility.cc#L27-L29) as well to run test propery.

**Comment by leyao-daily** [comment]
@googlebot I fixed it.

**Comment by leyao-daily** [comment]
@googlebot I consent.

**Comment by lum1n0us** [comment]
@googlebot I consent.

File: bazel/external/wamr.BUILD
Comment (PiotrSikora) [inline bazel/external/wamr.BUILD:7]: It looks that WAMR uses CMake. Could you use `rules_foreign_cc` to make it build? See this example: https://github.com/bazelbuild/rules_foreign_cc/blob/main/examples/cmake_hello_world_lib/shared/BUILD.bazel

Comment (leyao-daily) [reply bazel/external/wamr.BUILD:7]: Yes. Current building is using external dynamic lib, which is not recommended. I will then work on the static library build from wamr source github repo as the guide.

Comment (lum1n0us) [reply bazel/external/wamr.BUILD:7]: please be aware that the path of *shared_library* has been changed in the latest library package.

File: src/wamr/wamr.cc
Comment (lum1n0us) [inline src/wamr/wamr.cc:305]: WAMR currently tends to link *.WAMR* automatically. We have to remove all the manual linking code. L305 -> L371

**Comment by PiotrSikora** [comment]
@leyao-daily do you have any plans for integrating WAMR in Envoy (most of the integration is abstracted away and already done, but there are few places where support for the new runtime needs to be explicitly added)? We usually wait with merging until end-to-end tests pass.

**Comment by leyao-daily** [comment]
> @leyao-daily do you have any plans for integrating WAMR in Envoy (most of the integration is abstracted away and already done, but there are few places where support for the new runtime needs to be explicitly added)? We usually wait with merging until end-to-end tests pass.

Yes. We have plan to integrate WAMR in Envoy and the work is WIP. Should we create a PR or Issue for it? (Contribute to Envoy)

And could you give me some advices about the end-to-end tests? Is there any specific test cases for the runtime?

**Comment by PiotrSikora** [comment]
> Yes. We have plan to integrate WAMR in Envoy and the work is WIP. Should we create a PR or Issue for it? (Contribute to Envoy)

If you're working on it, then open PR once it's ready.
 
> And could you give me some advices about the end-to-end tests? Is there any specific test cases for the runtime?

Part of the PR in Envoy is going to be adding `--define wasm=wamr` (this is how other Wasm runtimes are supported there), in which case Envoy build WAMR instead of the default V8 runtime, and all existing tests will be executed using this runtime.

**Comment by leyao-daily** [comment]
Got it.

**Comment by PiotrSikora** [comment]
Please merge with `master` and apply changes from https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/152 when you get a chance. Thanks!

**Comment by leyao-daily** [comment]
> Please merge with `master` and apply changes from #152 when you get a chance. Thanks!

So, we need do the same change with wasmtime/v8 in wamr.cc?

**Comment by PiotrSikora** [comment]
> So, we need do the same change with wasmtime/v8 in wamr.cc?

Correct.

File: src/wamr/wamr.cc
Comment (PiotrSikora) [inline src/wamr/wamr.cc:162]: This is not how runtime with `Cloneable::CompiledBytecode` should behave. If you're creating a new instance from scratch, then you should return `Cloneable::NotCloneable` and not implement the `clone()` function.

**Review by PiotrSikora — CHANGES_REQUESTED** [review]

File: src/wamr/wamr.cc
Comment (leyao-daily) [reply src/wamr/wamr.cc]: Fixed

File: bazel/repositories.bzl
Comment (PiotrSikora) [inline bazel/repositories.bzl:59]: Please remove the commented out section. If you're using it for local development, you can use Bazel's `--override_repository=wamr=/source/wamr` instead.

File: bazel/repositories.bzl
Comment (PiotrSikora) [inline bazel/repositories.bzl:52]: Could you move this before `wasmtime`? You've added this in between `wasmtime` and `wasm_c_api` (used by Wasmtime), so it's easier to miss during updates.

File: BUILD
Comment (PiotrSikora) [inline BUILD:40]: Please revert those changes, since they replace Wasmtime with WAMR in CI tests.

We need to add support for multiple runtimes in tests, but that's not ready yet (patches welcome!).

File: WORKSPACE
Comment (PiotrSikora) [inline WORKSPACE:13]: Nit: add new line at the end.

File: bazel/external/wamr.BUILD
Comment (PiotrSikora) [inline bazel/external/wamr.BUILD:16]: I think you should remove this, Bazel is going to automatically set this to `Debug` when using `-c dbg`, and `Release` when using `-c fastbuild` and `-c opt`.

**Review by PiotrSikora — COMMENTED** [review]
Could you also pick up changes from #146? Otherwise, it looks good, thanks!

As mentioned before, I'm going to wait until Envoy PR is ready before merging.

**Comment by leyao-daily** [comment]
@PiotrSikora In envoy test, it will report based on the latest proxy-wasm
test/extensions/common/wasm/wasm_vm_test.cc:63:24: error: no member named 'getCustomSection' in 'proxy_wasm::WasmVm'
  EXPECT_TRUE(wasm_vm->getCustomSection("user").empty());

It seems that we remove getCustomSection in recent patch?

**Comment by PiotrSikora** [comment]
> @PiotrSikora In envoy test, it will report based on the latest proxy-wasm
> test/extensions/common/wasm/wasm_vm_test.cc:63:24: error: no member named 'getCustomSection' in 'proxy_wasm::WasmVm'
> EXPECT_TRUE(wasm_vm->getCustomSection("user").empty());
> 
> It seems that we remove getCustomSection in recent patch?

Yes, it was made private in #152.

File: bazel/repositories.bzl
Comment (leyao-daily) [reply bazel/repositories.bzl:44]: OK

File: bazel/external/wamr.BUILD
Comment (leyao-daily) [reply bazel/external/wamr.BUILD]: Fixed

File: BUILD
Comment (leyao-daily) [reply BUILD:40]: Fixed. And exclude the wamr*, because there will be some symbol conflict.

File: test/utility.cc
Comment (mathetake) [inline test/utility.cc:30]: where `WASM_WAMR` is defined in build targets..? I don't see any test is running correctly for WAMR so far..

Comment (leyao-daily) [reply test/utility.cc:30]: > where `WASM_WAMR` is defined in build targets..? I don't see any test is running correctly for WAMR so far..

In Proxy-wasm/Build, I locally set the target as WAMR_lib and exculde wasmtime_lib because of the potential symbol conficts. I just run the build test //test/... , where should I define the WAMS_WAMR

Comment (mathetake) [reply test/utility.cc:30]: OK, as long as Envoy test passes, we are fine then for now. 

As in https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/142#discussion_r613046016, we should fix CI here in another PR to test these runtimes (Wasmtime+WAMR) in GHA.

Comment (leyao-daily) [reply test/utility.cc:30]: Fine. Thanks a lot.

File: bazel/repositories.bzl
Comment (lum1n0us) [reply bazel/repositories.bzl]: seems can not assign a new *build_file* while using `--override_repository`. 
but, right, we should remove that debug information.

**Comment by leyao-daily** [comment]
@PiotrSikora Hi, I have created the PR in Envoy https://github.com/envoyproxy/envoy/pull/16057. And the bazel external repo points to proxy-wasm officer repo, so the WAMR check is faied in CI/CD (include tidy). It seems to be a loop conflict between the two project. Do you have any suggestion? Thanks (But locally I pass the testing using this PR repo address.)

**Comment by PiotrSikora** [comment]
> @PiotrSikora Hi, I have created the PR in Envoy [envoyproxy/envoy#16057](https://github.com/envoyproxy/envoy/pull/16057). And the bazel external repo points to proxy-wasm officer repo, so the WAMR check is faied in CI/CD (include tidy). It seems to be a loop conflict between the two project. Do you have any suggestion? Thanks (But locally I pass the testing using this PR repo address.)

In your PR in Envoy, you should point `proxy-wasm-cpp-host` at https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/142/commits/1afb82b7a506c0de538f2777e76f2f72728a7a23 (i.e. the latest commit from this PR).

**Comment by leyao-daily** [comment]
> > @PiotrSikora Hi, I have created the PR in Envoy [envoyproxy/envoy#16057](https://github.com/envoyproxy/envoy/pull/16057). And the bazel external repo points to proxy-wasm officer repo, so the WAMR check is faied in CI/CD (include tidy). It seems to be a loop conflict between the two project. Do you have any suggestion? Thanks (But locally I pass the testing using this PR repo address.)
> 
> In your PR in Envoy, you should point `proxy-wasm-cpp-host` at [1afb82b](https://github.com/proxy-wasm/proxy-wasm-cpp-host/commit/1afb82b7a506c0de538f2777e76f2f72728a7a23) (i.e. the latest commit from this PR).

Got it. Thanks

**Comment by leyao-daily** [comment]
@PiotrSikora Hi, we use the newest version of Proxy-wasm which delete 'getCustomSection' to common. And the wasm_vm_test cases in Envoy is the old version. So the CI/CD release check is always failed `https://dev.azure.com/cncf/envoy/_build/results?buildId=72844&view=logs&j=767be981-567e-57d8-68c3-2140ede0a0bd&t=1fb0b72f-6293-562d-a4d1-bb649223eaca&l=19266`.
We may need to update the wam vm test cases in Envoy?

**Comment by mathetake** [comment]
yes you must update. See https://github.com/envoyproxy/envoy/pull/16112/files#diff-37e9de86ad8b46c2b0bf492de5fa3d3dbf4165a58c5d0ba6486c923b8fff7189

File: bazel/repositories.bzl
Comment (PiotrSikora) [inline bazel/repositories.bzl:44]: This PR (https://github.com/bytecodealliance/wasm-micro-runtime/pull/618) isn't merged yet.

**Review by PiotrSikora — COMMENTED** [review]
Thanks! This looks good and passes tests in Envoy (tested locally).

Any plans for adding support for AOT via precompiled modules (already supported in V8 and WAVM runtimes)? It can be done in a separate PR.

File: bazel/repositories.bzl
Comment (lum1n0us) [reply bazel/repositories.bzl:44]: :laughing: We are working on JIT support now and gonna enable AOT in the next step.

Comment (PiotrSikora) [reply bazel/repositories.bzl:44]: Oh, nice!

Comment (lum1n0us) [reply bazel/repositories.bzl:44]: Does WAMR need to merge this PR (bytecodealliance/wasm-micro-runtime#618)  first ?

Comment (PiotrSikora) [reply bazel/repositories.bzl:44]: Ideally, yes. Otherwise, we're pulling some "random" commit that's not part of WAMR.

Comment (PiotrSikora) [reply bazel/repositories.bzl:44]: It looks that it was merged (by someone else?) in bytecodealliance/wasm-micro-runtime#622.

Could you update this PR?

Comment (PiotrSikora) [reply bazel/repositories.bzl:44]: Also, please merge `master` branch here to pickup other updates, otherwise your PR in Envoy won't work.

**Comment by lum1n0us** [comment]
sure. WAMR is preparing a release including all necessary enhancements.we will update the PR after the release is ready.

________________________________

File: bazel/repositories.bzl
Comment (leyao-daily) [reply bazel/repositories.bzl]: Update to master now.

File: bazel/external/wamr.BUILD
Comment (PiotrSikora) [inline bazel/external/wamr.BUILD:23]: Don't we need LLVM dependency like in Envoy?

File: bazel/external/wamr.BUILD
Comment (PiotrSikora) [inline bazel/external/wamr.BUILD:16]: Please match it to options used in Envoy. At the very least, please add:
```
"WAMR_BUILD_INTERP": "1",
"WAMR_BUILD_JIT": "0",
```
or does JIT work already?

File: src/wamr/wamr.cc
Comment (PiotrSikora) [inline src/wamr/wamr.cc:59]: This should return `"wamr"`, since Proxy-Wasm C++ Host is (in theory) proxy-agnostic.

The issue you've hit should be fixed in https://github.com/envoyproxy/envoy/pull/16263.

**Comment by leyao-daily** [comment]
Get it. And I am on holiday now, so I will update PR after May 6th. Thanks!

________________________________
发件人: Piotr Sikora ***@***.***>
发送时间: Sunday, May 2, 2021 6:29:10 AM
收件人: proxy-wasm/proxy-wasm-cpp-host ***@***.***>
抄送: Yao, Le ***@***.***>; Mention ***@***.***>
主题: Re: [proxy-wasm/proxy-wasm-cpp-host] Add Webassembly micro runtime support (#142)


@PiotrSikora commented on this pull request.

________________________________

In src/wamr/wamr.cc<https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/142#discussion_r624574714>:

> +  void *raw_func_;
+  WasmVm *vm_;
+};
+
+using HostFuncDataPtr = std::unique_ptr<HostFuncData>;
+
+wasm_engine_t *engine() {
+  static const auto engine = WasmEnginePtr(wasm_engine_new());
+  return engine.get();
+}
+
+class Wamr : public WasmVm {
+public:
+  Wamr() {}
+
+  std::string_view runtime() override { return "envoy.wasm.runtime.wamr"; }


This should return "wamr", since Proxy-Wasm C++ Host is (in theory) proxy-agnostic.

The issue you've hit should be fixed in envoyproxy/envoy#16263<https://github.com/envoyproxy/envoy/pull/16263>.

―

File: bazel/external/wamr.BUILD
Comment (lum1n0us) [reply bazel/external/wamr.BUILD:16]: :hand: just curious, I didn't find any WAVM related BUILD file. Does it mean *WAVM* can not be linked in this repo yet?

Comment (PiotrSikora) [reply bazel/external/wamr.BUILD:18]: Correct, currently you cannot build WAVM in this repo. We're slowly fixing this (see: https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/160), which is why it would be helpful if WAMR worked.

**Review by PiotrSikora — COMMENTED** [review]
@mathetake could you take a pass on this PR? I think it looks good, and tests in Envoy passed already (https://github.com/envoyproxy/envoy/pull/16057).

I think the only thing missing right now is the LLVM dependency, but perhaps we could merge it as-is to unblock merge in Envoy, and we could fix any build issues as part of https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/160.

What do you think?

**Comment by mathetake** [comment]
this looks good to me now and we could merge as-is since Envoy tests pass. And I will handle build/test issues here in https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/160 and subsequent PRs.

**Review by PiotrSikora — APPROVED** [review]
LGTM, thanks a lot! @leyao-daily @lum1n0us please let me know if this is ready to be merged or if you still want to make some changes, thanks!

**Comment by leyao-daily** [comment]
> LGTM, thanks a lot! @leyao-daily @lum1n0us please let me know if this is ready to be merged or if you still want to make some changes, thanks!

Cool! We think it is ready to be merged now. Thanks a lot.

**Comment by PiotrSikora** [comment]
Thanks!

**Pull request merged**
